File: Sample/java/src/main/java/net/nend/sample/java/nativead/NativeViewPagerActivity.java
package net.nend.sample.java.nativead;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import net.nend.android.NendAdNative;
import net.nend.android.NendAdNativeClient;
import net.nend.android.NendAdNativeListener;
import net.nend.android.NendAdNativeViewBinder;
import net.nend.sample.java.R;
import net.nend.sample.java.nativead.customviews.NativePagerFragment;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;

public class NativeViewPagerActivity extends AppCompatActivity implements NativePagerFragment.OnAdListener {

    private final String TAG = getClass().getSimpleName();
    private NendAdNativeViewBinder mBinder;
    private NendAdNativeClient mClient;
    // 広告を表示したポジションの一覧
    private List<Integer> mPositionList = new ArrayList<>();
    // 表示したポジションと広告を紐付けて保持
    private HashMap<Integer, NendAdNative> mLoadedAd = new HashMap<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.native_viewpager);

        ViewPager mViewPager = findViewById(R.id.pager);
        assert mViewPager != null;
        mViewPager.setAdapter(new CustomPagerAdapter(getSupportFragmentManager()));

        mBinder = new NendAdNativeViewBinder.Builder()
                .adImageId(R.id.ad_image)
                .logoImageId(R.id.logo_image)
                .titleId(R.id.ad_title)
                .contentId(R.id.ad_content)
                .prId(R.id.ad_pr, NendAdNative.AdvertisingExplicitly.SPONSORED)
                .promotionUrlId(R.id.ad_promotion_url)
                .promotionNameId(R.id.ad_promotion_name)
                .actionId(R.id.ad_action)
                .build();
        mClient = new NendAdNativeClient(this, 485520, "a88c0bcaa2646c4ef8b2b656fd38d6785762f2ff");
    }

    @Override
    public void onAdRequest(final View view, final int position) {
        if (mLoadedAd.containsKey(position)) {
            mLoadedAd.get(position).intoView(view, mBinder);
        } else {
            mClient.loadAd(new NendAdNativeClient.Callback() {
                @Override
                public void onSuccess(final NendAdNative nendAdNative) {
                    Log.i(TAG, "広告取得成功");
                    mLoadedAd.put(position, nendAdNative);
                    mPositionList.add(position);
                    final NendAdNative adNative = mLoadedAd.get(position);
                    if (adNative != null) {
                        adNative.intoView(view, mBinder);
                        adNative.setNendAdNativeListener(new NendAdNativeListener() {
                            @Override
                            public void onImpression(@NonNull NendAdNative nendAdNative) {
                                Log.i(TAG, "onImpression");
                            }

                            @Override
                            public void onClickAd(@NonNull NendAdNative nendAdNative) {
                                Log.i(TAG, "onClickAd");
                            }

                            @Override
                            public void onClickInformation(@NonNull NendAdNative nendAdNative) {
                                Log.i(TAG, "onClickInformation");
                            }
                        });
                    }
                }

                @Override
                public void onFailure(NendAdNativeClient.NendError nendError) {
                    Log.i(TAG, "広告取得失敗 " + nendError.getMessage());
                    // すでに取得済みの広告があればランダムで表示
                    if (!mLoadedAd.isEmpty()) {
                        Collections.shuffle(mPositionList);
                        Objects.requireNonNull(mLoadedAd.get(mPositionList.get(0))).intoView(view, mBinder);
                    }
                }
            });
        }
    }

    public static class CustomPagerAdapter extends FragmentPagerAdapter {

        public CustomPagerAdapter(FragmentManager fm) {
            super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            NativePagerFragment fragment = new NativePagerFragment();
            Bundle extras = new Bundle();
            extras.putInt("position", position);
            fragment.setArguments(extras);
            return fragment;
        }

        @Override
        public int getCount() {
            return 5;
        }
    }
}

File: Sample/kotlin/src/main/java/net/nend/sample/kotlin/nativead/NativeViewPagerActivity.kt
package net.nend.sample.kotlin.nativead

import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import net.nend.android.NendAdNative
import net.nend.android.NendAdNativeClient
import net.nend.android.NendAdNativeListener
import net.nend.android.NendAdNativeViewBinder
import net.nend.sample.kotlin.R
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_API_KEY_LARGE_WIDE
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_SPOT_ID_LARGE_WIDE

class NativeViewPagerActivity : AppCompatActivity(), NativePagerFragment.OnAdListener {

    private lateinit var binder: NendAdNativeViewBinder
    private lateinit var client: NendAdNativeClient
    // 広告を表示したポジションの一覧
    private val positionList = mutableListOf<Int>()
    // 表示したポジションと広告を紐付けて保持
    private val loadedAd = mutableMapOf<Int, NendAdNative>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.native_viewpager)

        val viewPager = findViewById<ViewPager>(R.id.pager)
        viewPager.adapter = CustomPagerAdapter(supportFragmentManager)

        binder = NendAdNativeViewBinder.Builder()
                .adImageId(R.id.ad_image)
                .logoImageId(R.id.logo_image)
                .titleId(R.id.ad_title)
                .contentId(R.id.ad_content)
                .prId(R.id.ad_pr, NendAdNative.AdvertisingExplicitly.SPONSORED)
                .promotionUrlId(R.id.ad_promotion_url)
                .promotionNameId(R.id.ad_promotion_name)
                .actionId(R.id.ad_action)
                .build()
        client = NendAdNativeClient(this, NATIVE_SPOT_ID_LARGE_WIDE, NATIVE_API_KEY_LARGE_WIDE)
    }

    override fun onAdRequest(view: View, position: Int) {
        if (loadedAd.containsKey(position)) {
            loadedAd[position]?.intoView(view, binder)
        } else {
            client.loadAd(object : NendAdNativeClient.Callback {
                override fun onSuccess(nendAdNative: NendAdNative) {
                    Log.i(TAG, "広告取得成功")
                    positionList.add(position)
                    nendAdNative.run {
                        intoView(view, binder)
                        setNendAdNativeListener(object : NendAdNativeListener {
                            override fun onImpression(ad: NendAdNative) {
                                Log.i(TAG, "onImpression")
                            }

                            override fun onClickAd(ad: NendAdNative) {
                                Log.i(TAG, "onClickAd")
                            }

                            override fun onClickInformation(ad: NendAdNative) {
                                Log.i(TAG, "onClickInformation")
                            }
                        })
                        loadedAd[position] = this
                    }
                }

                override fun onFailure(nendError: NendAdNativeClient.NendError) {
                    Log.i(TAG, "広告取得失敗: ${nendError.message} ")
                    // すでに取得済みの広告があればランダムで表示
                    if (loadedAd.isNotEmpty()) {
                        positionList.shuffle()
                        loadedAd[positionList[0]]?.intoView(view, binder)
                    }
                }
            })
        }
    }

    inner class CustomPagerAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
        override fun getItem(position: Int): Fragment {
            return NativePagerFragment().apply {
                arguments = Bundle().apply {
                    putInt("position", position)
                }
            }
        }

        override fun getCount() = 5
    }

    companion object {
        private val TAG = NativeViewPagerActivity::class.java.simpleName
    }
}